src/txrequest.cpp

```diff
@@ -0,0 +1,685 @@
+// Copyright (c) 2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <txrequest.h>
+
+#include <crypto/siphash.h>
+#include <net.h>
+#include <primitives/transaction.h>
+#include <random.h>
+#include <uint256.h>
+#include <util/memory.h>
+
+#include <boost/multi_index_container.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+
+#include <chrono>
+#include <unordered_map>
+#include <utility>
+
+#include <assert.h>
+
+namespace {
+
+/** The various states a (txhash,peer) pair can be in.
+ *
+ * Note that CANDIDATE is split up into 3 substates (DELAYED, BEST, READY), allowing more efficient implementation.
+ * Also note that the sorting order of EntryTxHash relies on the specific order of values in this enum.
+ *
+ * Expected behaviour is:
+ *   - When first announced by a peer, the state is CANDIDATE_DELAYED until reqtime is reached.
+ *   - Announcemnets that have reached their reqtime but not been requested will be either CANDIDATE_READY or
+ *     CANDIDATE_BEST
+ *   - When requested, an announcement will be in state REQUESTED until expiry is reached.
+ *   - If expiry is reached, or the peer replies to the request (either with NOTFOUND or the tx), the state becomes
+ *     COMPLETED
+ */
+enum class State : uint8_t {
+    /** A CANDIDATE entry whose reqtime is in the future. */
+    CANDIDATE_DELAYED,
+    /** The best CANDIDATE for a given txhash; only if there is no REQUESTED entry already for that txhash.
+     *  The CANDIDATE_BEST is the lowest-priority entry among all CANDIDATE_READY (and _BEST) ones for that txhash. */
+    CANDIDATE_BEST,
+    /** A REQUESTED entry. */
+    REQUESTED,
+    /** A CANDIDATE entry that's not CANDIDATE_DELAYED or CANDIDATE_BEST. */
+    CANDIDATE_READY,
+    /** A COMPLETED entry. */
+    COMPLETED,
+
+    /** An invalid State value that's larger than all valid ones. */
+    TOO_LARGE,
+};
+
+/** An announcement entry. This is the data we track for each txid or wtxid that is announced to us. */
+struct Entry {
```

Comment:
Are "entry" and "announcement" synonymous? If so, would it make sense to rename `Entry` to `Announcement` and drop the "entry" terminology?

Reply:
+1 was wondering the same

Reply:
Done, renamed a few things:
* Entry -> Announcement
* Entry{TxHash,Peer,Time} -> By{TxHash,Peer,Time}View
* Entry{TxHash,Peer,Time}Extractor -> By{TxHash,Peer,Time}ViewExtractor

No more "entr" anywhere in txrequest (I did not make the same change in the fuzz test, though).